File: Bai_58/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Item {
    int weight;
    int value;
};

bool compare(Item a, Item b) {
    return (a.value * 1.0 / a.weight) > (b.value * 1.0 / b.weight);
}

int main() {
    freopen("SAMPLE.INP", "r", stdin);
    freopen("SAMPLE.OUT", "w", stdout);

    int N, M;
    cin >> N >> M;

    vector<Item> items(N);
    for (int i = 0; i < N; ++i) {
        cin >> items[i].weight >> items[i].value;
    }

    vector<int> capacities(M);
    for (int i = 0; i < M; ++i) {
        cin >> capacities[i];
    }

    sort(items.begin(), items.end(), compare);

    long long totalValue = 0;
    for (int i = 0; i < N && M > 0; ++i) {
        for (int j = 0; j < M; ++j) {
            if (capacities[j] >= items[i].weight) {
                totalValue += items[i].value;
                capacities[j] -= items[i].weight;
                break;
            }
        }
    }

    cout << totalValue << endl;

    return 0;
}

File: Bai_58/main.py

with open("E:\Project Visual Studio Code\Summary_exercises\Bai_55\SAMPLE.INP", "r") as f_in, open("SAMPLE.OUT", "w") as f_out:
    N, M = map(int, f_in.readline().split())
    
    items = []
    for _ in range(N):
        weight, value = map(int, f_in.readline().split())
        items.append((weight, value))

    capacities = [int(f_in.readline()) for _ in range(M)]

    items.sort(key=lambda x: x[1] / x[0], reverse=True)

    total_value = 0
    for weight, value in items:
        for i in range(M):
            if capacities[i] >= weight:
                total_value += value
                capacities[i] -= weight
                break

    f_out.write(str(total_value) + "\n")